src/coap_blockwise.c

```diff
@@ -434,7 +436,12 @@ static void on_block_rcvd(struct golioth_client *client,
 
     if (is_last || GOLIOTH_OK != status)
     {
-        ctx->end_cb(client, status, coap_rsp_code, path, ctx->block_idx, ctx->callback_arg);
+        /* If we are ending due to an error on the first block of a response to a post,
+           then the error was already sent to the application via the set_block callback. */
+        if (GOLIOTH_OK == status || !ctx->post_response || ctx->block_idx > 0)
+        {
+            ctx->end_cb(client, status, coap_rsp_code, path, ctx->block_idx, ctx->callback_arg);
+        }
 
         golioth_sys_free(ctx);
     }
```

Comment:
Pulling out `if(GOLIOTH_OK != status && ctx->post_response && ctx->block_idx == 0)` into a top-level conditional (where we only free and don't call `ctx->end_cb` ) rather than nesting feels slightly easier to digest to me, but open to the structure you have here as well.